syntax = "proto3";

package amanda;

import "value.proto";

message DataTypeDef {
  string namespace = 1;
  string name = 2;
  map<string, Value> attrs = 3;
  Value raw = 4;
}

message PortDef {
  string name = 1;
  DataTypeDef type = 2;
}

message EdgeDef {
  string src_op = 1;
  string src_port = 2;
  string dst_op = 3;
  string dst_port = 4;
  map<string, Value> attrs = 5;
}

message NodeDef {
  string name = 1;
  string type = 2;
  string namespace = 3;
  repeated PortDef input_ports = 4;
  repeated PortDef output_ports = 5;
  repeated NodeDef ops = 6;
  repeated EdgeDef edges = 7;
  map<string, Value> attrs = 8;
}
